### ISSUE TITLE
BinaryHeapQuickRemovals O(log(n)) runtime may not be correct? 

### ISSUE THREAD

**Issue by Jun711** [williamfiset/Algorithms#165]
Hi William, thank you for your data structure videos. I have a question regarding your implementation of BinaryHeapQuickRemovals.

```
// This map keeps track of the possible indices a particular
// node value is found in the heap. Having this mapping lets
// us have O(log(n)) removals and O(1) element containment check
// at the cost of some additional space and minor overhead
private Map<T, TreeSet<Integer>> map = new HashMap<>();
```
![Screen Shot 2020-04-17 at 4 11 30 PM](https://user-images.githubusercontent.com/6880939/79620828-1c9aac00-80c6-11ea-875e-47ec102c28c8.png)

When we map one value to multiple positions using a TreeSet, runtime complexity of removing elements from priority queue is probably not O(log (n)). 

It seems to be O (log (n) * log (n)) because TreeSet remove operation takes O(log (n)) time if I am not mistaken. However, I think it would be O(log (n)) if we use a HashSet instead of a TreeSet as HashSet has O(1) time for removal. 

Using your PQ removal example, when we remove 2 from the PQ (refer to the example in your [video](https://youtu.be/eVq8CmoC1x8?list=PLDV1Zeh2NRsB6SWUrDFW2RmDotAfPbeHu&t=776)), 
1) 2 is swapped with 11. 11 is at the root of the PQ now. The operation to remove position 7 from node 2's TreeSet takes O(log (n))

2) when 11 is bubbled down, it is swapped with 2. 2 is at the root of the PQ now. The operation to remove position 0 from node 2's TreeSet takes O(log (n))

**Comment by williamfiset** [comment]
I think what you're describing is a O(log(n)) + O(log(n)) operation  rather than a O(log(n)) * O(log(n)) operation?

At the time when I made this video, I was still not aware of a data structure called an Indexed Priority Queue which is a variation of a priority queue that supports removals in O(log(n)) in a non-hacky way. I made a video on [indexed priority queues](https://www.youtube.com/watch?v=DT8xZ0Uf8wo) if you're interested. Basically, rather than mapping values to indexes, the IPQ maintains a (key, value) pairing for each node so that nodes can easily be accessed.
